src/vivarium/framework/randomness.py

```diff
@@ -314,21 +329,46 @@ def choice(
         If any row in `p` contains `RESIDUAL_CHOICE` and the remaining
         weights in the row are not normalized or any row of `p` contains
         more than one reference to `RESIDUAL_CHOICE`.
-
     """
-    # Convert p to normalized probabilities broadcasted over index.
-    p = (
-        _set_residual_probability(_normalize_shape(p, index))
-        if p is not None
-        else np.ones((len(index), len(choices)))
-    )
-    p = p / p.sum(axis=1, keepdims=True)
 
-    draw = random(key, index, index_map)
+    if (len(index) * len(choices) > large_N) & (p is not None):
+        warnings.warn(
```

Comment:
Should use the `loguru` `logger`. That is how we do logging elsewhere in vivarium